src/types/mappingproxy.rs

```diff
@@ -0,0 +1,614 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+
+use super::PyMapping;
+use crate::err::PyResult;
+use crate::ffi_ptr_ext::FfiPtrExt;
+use crate::instance::Bound;
+use crate::types::any::PyAnyMethods;
+use crate::types::{PyAny, PyIterator, PySequence};
+use crate::{ffi, Python};
+
+/// Represents a Python `mappingproxy`.
+#[repr(transparent)]
+pub struct PyMappingProxy(PyAny);
+
+pyobject_native_type_core!(
+    PyMappingProxy,
+    pyobject_native_static_type_object!(ffi::PyDictProxy_Type),
+    #checkfunction=ffi::PyDictProxy_Check
+);
+
+impl PyMappingProxy {
+    /// Creates a mappingproxy from an object.
+    pub fn new<'py>(
+        py: Python<'py>,
+        elements: &Bound<'py, PyMapping>,
+    ) -> Bound<'py, PyMappingProxy> {
+        unsafe {
+            ffi::PyDictProxy_New(elements.as_ptr())
+                .assume_owned(py)
+                .downcast_into_unchecked()
+        }
+    }
+}
+
+/// Implementation of functionality for [`PyMappingProxy`].
+///
+/// These methods are defined for the `Bound<'py, PyMappingProxy>` smart pointer, so to use method call
+/// syntax these methods are separated into a trait, because stable Rust does not yet support
+/// `arbitrary_self_types`.
+#[doc(alias = "PyMappingProxy")]
+pub trait PyMappingProxyMethods<'py, 'a>: crate::sealed::Sealed {
+    /// Returns a new mappingproxy that contains the same key-value pairs as self.
+    ///
+    /// This is equivalent to the Python expression `self.copy()`.
+    fn copy(&self) -> PyResult<Bound<'_, PyMappingProxy>>;
+
+    /// Checks if the mappingproxy is empty, i.e. `len(self) == 0`.
+    fn is_empty(&self) -> PyResult<bool>;
+
+    /// Returns a sequence containing all keys in the mapping.
+    fn keys(&self) -> PyResult<Bound<'py, PySequence>>;
+
+    /// Returns a sequence containing all values in the mapping.
+    fn values(&self) -> PyResult<Bound<'py, PySequence>>;
+
+    /// Returns a sequence of tuples of all (key, value) pairs in the mapping.
+    fn items(&self) -> PyResult<Bound<'py, PySequence>>;
+
+    /// Returns `self` cast as a `PyMapping`.
+    fn as_mapping(&self) -> &Bound<'py, PyMapping>;
+
+    /// Takes an object and returns an iterator for it. Returns an error if the object is not
+    /// iterable.
+    fn try_iter(&'a self) -> PyResult<BoundMappingProxyIterator<'py, 'a>>;
+}
+
+impl<'py, 'a> PyMappingProxyMethods<'py, 'a> for Bound<'py, PyMappingProxy> {
+    fn copy(&self) -> PyResult<Bound<'_, PyMappingProxy>> {
+        let res = self.call_method0("copy")?;
+        unsafe { Ok(res.downcast_into_unchecked::<PyMappingProxy>()) }
+    }
```

Comment:
I think the type here is incorrect, it looks like `.copy()` returns a copy of the underlying mapping, perhaps?

```python
>>> type(types.MappingProxyType({}).copy())
<class 'dict'>
```

Reply:
I changed this to return `PyMapping` instead.